python/lsst/image_cutout_backend/_image_cutout_backend.py

```diff
@@ -0,0 +1,412 @@
+    """
+
+    def mask(self, name: str = "STENCIL") -> None:
+        """Set the bitmask to show the approximate coverage of nonrectangular
+        stencils.
+
+        Parameters
+        ----------
+        name : `str`, optional
+            Name of the mask plane to add and set.
+
+        Notes
+        -----
+        This method modifies `cutout` in place if it is a `Mask`,
+        `MaskedImage`, or `Exposure`.  It does nothing if `cutout` is an
+        `Image`.
+        """
+        if isinstance(self.cutout, Exposure):
+            mask = self.cutout.mask
+        elif isinstance(self.cutout, MaskedImage):
+            mask = self.cutout.mask
+        elif isinstance(self.cutout, Mask):
+            mask = self.cutout
+        else:
+            return
+        mask.addMaskPlane(name)
+        bits = mask.getPlaneBitMask(name)
+        self.pixel_stencil.set_mask(mask, bits)
+
+    def write_fits(self, path: str) -> None:
+        """Write the cutout to a FITS file.
+
+        Parameters
+        ----------
+        path : `str`
+            Local path to the file.
+
+        Notes
+        -----
+        If ``cutout`` is an `Exposure`, this will merge `metadata` into the
+        cutout's attached metadata.  In other cases, `metadata` is written
+        to the primary header without modifying `cutout`.
+        """
+        if isinstance(self.cutout, Exposure):
+            self.cutout.getMetadata().update(self.metadata)
+            self.cutout.writeFits(path)
+        else:
+            self.cutout.writeFits(path, metadata=self.metadata)
+
+
+class ImageCutoutBackend:
+    """High-level interface to the image cutout backend.
+
+    Parameters
+    ----------
+    butler : `Butler`
+        Butler that image cutouts are read from.
+    projection_finder : `ProjectionObject`
+        Object that obtains the WCS and bounding box for butler datasets of
+        different types.  May include caches.
+    output_root : convertible to `ResourcePath`
+        Root of output file URIs.  This will be combined with the originating
+        dataset's UUID and an encoding of the stencil to form the complete URI.
+    filename_template : `str`, optional
+        Template for output file names, not including the ``output_root``
+        prefix or the file type extension.  This is a `str.format` template
+        with named substitutions.  These must include ``stencil`` (the sky
+        stencil fingerprint in hex), and a unique set of fields from ``ref``
+        (the origin `DatasetRef`), such as ``ref.id`` or a combination of the
+        dataset type, `~lsst.daf.butler.CollectionType.RUN`` collection, and
+        data ID, but ``ref.dataId`` may not be expanded.
+    temporary_root : convertible to `ResourcePath`, optional
+        Local filesystem root to write files to before they are transferred to
+        ``output_root`` (passed as the prefix argument to
+        `ResourcePath.temporary_uri`).
```

Comment:
Why is this required? I'd rather not write the files to local disk because then I have to worry about available space on the local disk.

Reply:
It's because all of our low-level FITS I/O code goes through the cfitsio C library, and that usually operates on regular C file pointers, so we don't have a direct object store write (or read) for any of these types; this local temporary file stuff is what happens behind Butler for these objects, too.  It should be possible to tell cfitsio to write to a contiguous block of memory that we've allocated for it instead, and then send those bytes directly to an object store, but we'll need to do some plumbing in obs_base.  For large cutouts, it also means you've traded a free filesystem space question, if that matters (though maybe the filesystem one is actually worse because the temporaries aren't freed up as reliably?).

Reply:
The problem isn't so much cleaning up temporary space as that it's often easier to allocate memory than disk in cloud environments. But it sounds like it's inherent in the libraries we're using. I don't think it's worth going to a lot of effort to change until we see whether it's a problem.